v-next/hardhat/test/internal/builtin-plugins/solidity/build-system/integration/target-sources.ts

```diff
@@ -0,0 +1,258 @@
+import assert from "node:assert/strict";
+import path from "node:path";
+import { describe, it } from "node:test";
+
+import {
+  exists,
+  readJsonFile,
+  writeUtf8File,
+} from "@nomicfoundation/hardhat-utils/fs";
+
+import { useTestProjectTemplate } from "../resolver/helpers.js";
+
+const basicProjectTemplate = {
+  name: "test",
+  version: "1.0.0",
+  files: {
+    "contracts/Foo.sol": `// SPDX-License-Identifier: UNLICENSED \n pragma solidity ^0.8.0; contract Foo {}`,
+    "contracts/Foo.t.sol": `
+      // SPDX-License-Identifier: UNLICENSED
+      pragma solidity ^0.8.0;
+
+      import {Foo} from "./Foo.sol";
+
+      contract FooTest {
+        Foo foo;
+
+        function setUp() public {
+          foo = new Foo();
+        }
+
+        function test_Assertion() public view {
+          require(1 == 1, "test assertion");
+        }
+      }
+    `,
+    "test/OtherFooTest.sol": `
+      // SPDX-License-Identifier: UNLICENSED
+      pragma solidity ^0.8.0;
+
+      import {Foo} from "../contracts/Foo.sol";
+
+      contract OtherFooTest {
+        Foo foo;
+
+        function setUp() public {
+          foo = new Foo();
+        }
+
+        function test_Assertion() public view {
+          require(1 == 1, "test assertion");
+        }
+      }
+    `,
+  },
+};
+
+describe("build system - build task - behavior on target sources", function () {
+  describe("compiling contracts", function () {
+    it("compiles and generates artifacts for all contracts, and no artifacts for tests", async () => {
+      await using project = await useTestProjectTemplate(basicProjectTemplate);
+      const hre = await project.getHRE();
+
+      await hre.tasks.getTask("build").run(); // not specifying targetSources, testing it defaults to 'contracts'
+
+      const contractsArtifactsPath =
+        await hre.solidity.getArtifactsDirectory("contracts");
+
+      // Read artifact file (asserts it exists)
+      const artifact: any = await readJsonFile(
+        path.join(contractsArtifactsPath, "contracts", `Foo.sol`, `Foo.json`),
+      );
+
+      // Assert build info file exists
+      const buildInfoPath = path.join(
+        contractsArtifactsPath,
+        "build-info",
+        `${artifact.buildInfoId}.json`,
+      );
+      assert.equal(await exists(buildInfoPath), true);
```

Comment:
This test doesn't seem complete to me. For example: maybe the test in the `test/` directory is not being compiled, but the `.t.sol` one in `contracts` is.

Easy fix: load the build info and check that it only has a single source.